[UI] Dynamic Survey View - Ask teacher for class info #4612

Annelein opened the pull request:
Fixes #4612 

Survey view:
- Takes a survey_id, a description and questions
- 'Remind me later' sets 'skip' in survey to todays date, so you could modify the survey to pop up tomorrow/next week
- 'Don't ask again' sets 'skip' in survey to True, so you could choose to never show the survey again
- Option to only show unanswered questions

Survey table:
- Survey_id is: `survey_name + '_' + response_id` (for ex for the class survey: 'class_*class_id*')
- Responses
- Skip option

![Screenshot 2023-11-16 at 15 50 06](https://github.com/hedyorg/hedy/assets/48122190/2937268a-7009-482a-b3c4-f7fee27416c0)

jpelay commented:
Hi, @Annelein! I reckon I helped you out a bit on Discord about the HTMX issue, is there anything else you need help?

Annelein commented:
> Hi, @Annelein! I reckon I helped you out a bit on Discord about the HTMX issue, is there anything else you need help?

Yes you definitely helped! But working on some Hour of Code issues right now, after that I will get back at it and will get in touch with you if I need any help, thanks!!

jpelay commented:
> > Hi, @Annelein! I reckon I helped you out a bit on Discord about the HTMX issue, is there anything else you need help?
> 
> Yes you definitely helped! But working on some Hour of Code issues right now, after that I will get back at it and will get in touch with you if I need any help, thanks!!

Great! Since this is a PR still in progress I'll set it as a draft :smile: when this is done you can set it as ready for review.

Felienne commented on translations/en/LC_MESSAGES/messages.po:
```diff
@@ -1,168 +1,86 @@
-"You tried to use the variable {name} on line {access_line_number}, but "
-"you set it on line {definition_line_number}. Set a variable before using "
-"it."
+msgstr "You tried to use the variable {name} on line {access_line_number}, but you set it on line {definition_line_number}. Set a variable before using it."
```
I am not sure why all these weird changes happen, I think @hasan-sh had a similar issue, do you remember that?

Felienne commented on translations/en/LC_MESSAGES/messages.po:
```diff
@@ -364,6 +262,16 @@ msgstr "Please enter the name of the new class"
+"What is the age range in your class?\n"
+"What is the spoken language in your class?\n"
+"What is the gender balance in your class?\n"
+"What distingueashes your students from others?"
```
```suggestion
"What distinguishes your students from others?"
```

Felienne commented:
Maybe --omit-header should be removed here?

<img width="867" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/d5450fdd-8d0e-4d84-b4e8-f886a0144ef7">

Tagging @jpelay who might also know!

jpelay commented on .run/Python tests in test_level.run.xml:
Why are we modifying this file? Seems like this is IDE specific?

jpelay commented:
> Maybe --omit-header should be removed here?
> 
> <img alt="image" width="867" src="https://user-images.githubusercontent.com/1003685/283514806-d5450fdd-8d0e-4d84-b4e8-f886a0144ef7.png">
> Tagging @jpelay who might also know!

The command works fine for me if I use the one in the wiki

Felienne commented:
> > Maybe --omit-header should be removed here?
> > <img alt="image" width="867" src="https://user-images.githubusercontent.com/1003685/283514806-d5450fdd-8d0e-4d84-b4e8-f886a0144ef7.png">
> > Tagging @jpelay who might also know!
> 
> The command works fine for me if I use the one in the wiki

Weird! But this must not be merged, it will upset Weblate without the header and with the new changes.

jpelay commented:
Can we make the survey a tad more responsive?

![image](https://github.com/hedyorg/hedy/assets/45865185/e125d24b-0a54-4fc4-9de7-e6b2df6c358a)

Annelein commented:
> Can we make the survey a tad more responsive?
> 
> ![image](https://user-images.githubusercontent.com/45865185/283520791-e125d24b-0a54-4fc4-9de7-e6b2df6c358a.png)

yeah I askes here https://github.com/hedyorg/hedy/pull/4711#issuecomment-1814607155
not sure how to do it

jpelay commented:
> Weird! But this must not be merged, it will upset Weblate without the header and with the new changes.

Sure! @Annelein can you try generating the .po files again, if it doesn't work I can generate them on my computer.

Annelein commented:
> > Weird! But this must not be merged, it will upset Weblate without the header and with the new changes.
> 
> Sure! @Annelein can you try generating the .po files again, if it doesn't work I can generate them on my computer.

done! let me know if this works, also it somehow rendered these:

```
#~ msgid "class_survey_questions"
#~ msgstr ""

#~ msgid "survey-test"
#~ msgstr ""
```
not sure how to solve that?

jpelay commented:
> yeah I askes here [#4711 (comment)](https://github.com/hedyorg/hedy/pull/4711#issuecomment-1814607155) not sure how to do it

Oh sorry! Instead of using a predefined `width` on the style tag of thad div, you can try using Tailwind styles. Here is the documentation to define the width: https://tailwindcss.com/docs/width

With tailwind you can also define media queries, and change the class applied to the element depending on the screen size. You can find that here  https://tailwindcss.com/docs/width#breakpoints-and-media-queries

Annelein commented:
> > yeah I askes here [#4711 (comment)](https://github.com/hedyorg/hedy/pull/4711#issuecomment-1814607155) not sure how to do it
> 
> Oh sorry! Instead of using a predefined `width` on the style tag of thad div, you can try using Tailwind styles. Here is the documentation to define the width: https://tailwindcss.com/docs/width
> 
> With tailwind you can also define media queries, and change the class applied to the element depending on the screen size. You can find that here https://tailwindcss.com/docs/width#breakpoints-and-media-queries

thank you!!

jpelay commented:
> > > Weird! But this must not be merged, it will upset Weblate without the header and with the new changes.
> > 
> > 
> > Sure! @Annelein can you try generating the .po files again, if it doesn't work I can generate them on my computer.
> 
> done! let me know if this works, also it somehow rendered these:
> 
> ```
> #~ msgid "class_survey_questions"
> #~ msgstr ""
> 
> #~ msgid "survey-test"
> #~ msgstr ""
> ```
> 
> not sure how to solve that?

This happens because you're not using those labels anymore, perhaps you commented them out by accident or removed them?

jpelay commented on translations/bg/LC_MESSAGES/messages.po:
```diff
@@ -1,167 +1,104 @@
-"The {at} command may not be used from level 16 onward. You can use square"
-" brackets to use an element from a list, for example `friends[i]`, "
-"`lucky_numbers[{random}]`."
+msgstr "The {at} command may not be used from level 16 onward. You can use square brackets to use an element from a list, for example `friends[i]`, `lucky_numbers[{random}]`."
 
```
It happened again :c Can you tell the exact command you're using, @Annelein?

jpelay replied:
Nevermind, I reproduced the issue! Will try figure this one out. Don't worry about this and continue working on the actual implementation.

Annelein replied:
```
pybabel extract -F babel.cfg -o messages.pot . --no-location --sort-output --omit-header
pybabel update -i messages.pot -d translations -N --no-wrap --omit-header
pybabel compile -f -d translations
```

Annelein replied:
thank you!

jpelay replied:
Yeah, your command is working as we expected, it's weird that main has the lines wrapped and is causing these conflicts

Felienne commented:
Hi @Annelein!

There are now a lot of conflicts in the py, some of which I can't solve. Could you do that?

Then could you try:

* In the po files, could you accept changes from main
* then: `pybabel extract -F babel.cfg -o messages.pot . --no-location --sort-output`
* then: `pybabel update -i messages.pot -d translations -N`

I suspect this will solve the weird changes in the po files (and if so we need to update the docs).

Annelein commented:
> Hi @Annelein!
> 
> There are now a lot of conflicts in the py, some of which I can't solve. Could you do that?
> 
> Then could you try:
> 
> * In the po files, could you accept changes from main
> * then: `pybabel extract -F babel.cfg -o messages.pot . --no-location --sort-output`
> * then: `pybabel update -i messages.pot -d translations -N`
> 
> I suspect this will solve the weird changes in the po files (and if so we need to update the docs).

I've merged and pushed, is everything like it should now?

Felienne commented:
for the po-files, yes:

<img width="890" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/9676fe46-395c-41c1-9a6e-029dfce3220f">

Thanks for bearing with us!

Maybe @jpelay has other comments on the PR, I will let him approve it if he is happy.

jpelay commented on translations/bg/LC_MESSAGES/messages.po:
```diff
@@ -2447,3 +2484,13 @@ msgstr "Your program"
 #~ msgid "title_class statistics"
 #~ msgstr "My statistics"
+<<<<<<< HEAD
+
```
Ooops we probably want to remove this :)

Annelein commented:
There were still some merge issues, I fixed all of them but `skip_survey` `later_survey` and `submit_survey` were missing in the translation files. So I've updated them but now some tag translations are also moving around and I am not sure what to do with it @jpelay @Felienne

Felienne commented on translations/bn/LC_MESSAGES/messages.po:
```diff
@@ -2077,18 +2089,6 @@ msgstr "Subscribe to the newsletter"
 msgid "surname"
 msgstr "Username"
 
-#, fuzzy
```
Tagging @hasan-sh for this one. Are these tags still in use or are they deleted correctly?

jpelay replied:
I made a quick search on main, and these tags are not included in any HTML or python file, so maybe it was correct?

Annelein replied:
I don't get why these are removed...

Felienne replied:
I just started a discusion in Discord, I think I understand why, but this is in any case not a problem with the PR @Annelein

jpelay commented:
I tested it across several languages and it works great!! Amazing work @Annelein! Let's wait for @hasan-sh input on the missing labels issue. We now just have to make the tests pass and it'll be ready to go! :smile:

jpelay commented on website/surveys.py:
```diff
@@ -0,0 +1,51 @@
+        for question, answer in request.form.items():
+            if not answer:
+                survey_done = False
+            responses[question] = answer
```
This is also a good point! I didn't try switching languages between sessions. Maybe this will also prevent errors with respect to translators messing with the format.

Annelein replied:
done!

Annelein commented on templates/htmx-survey.html:
```diff
@@ -0,0 +1,42 @@
+                hx-post="/surveys/skip-survey/{{ survey_id }}"
+                hx-swap="innerHTML"
+                _="on click toggle .hidden on #survey">
+                <u>{{_('survey_skip')}}</u></button>
```
I've changed the text to 'Remind me tomorrow' is that fine?

Annelein commented on website/database.py:
```diff
@@ -663,6 +673,21 @@ def update_class_data(self, id, class_data):
+        SURVEYS.update({"id": id}, {"skip": True})
+
+    def add_remind_later_survey(self, id):
+        SURVEYS.update({"id": id}, {"skip": date.today().isoformat()})
```
I've made skipping the survey generic. This survey will be shown again tomorrow, such as checked in `class_survey in for_teachers.py`. And the text on the button is also loaded there, which now says, 'Remind me tomorrow'. So if we want a survey in the future that is only shown once per week. It would be specified in for example `class_survey in for_teachers.py`. Is that fine?

hasan-sh replied:
Wonderful!

Annelein commented on website/surveys.py:
```diff
@@ -0,0 +1,51 @@
+from .website_module import WebsiteModule, route
+
+
+class SurveysModule(WebsiteModule):
```
thanks a lot! I thought of that as well, but it won't be generic for if we want to survey students for example

hasan-sh replied:
true that, we may or may not require users to be logged in as well. Let's see what happens in the future.

jpelay commented on website/__init__.py:
I think the deletion of this file is what is causing the validation problems in the Unit test job, can you add it again and see if it solves the issue?

Felienne commented:
Woooohooo, the unit tests are passing now! Leaving the final approval to @hasan-sh since he still has a "request changes" review pending.

Annelein commented:
@Felienne @jpelay only the English and Arabic translations are there, all the other ones are gone, also the Dutch ones..?

Felienne commented:
> @Felienne @jpelay only the English and Arabic translations are there, all the other ones are gone, also the Dutch ones..?

Ah yes that happened when I was fixing merge conflicts, sorry. Feel free to add them again, but if we run into conflicts we can also add them later in a next PR.

hasan-sh commented on website/for_teachers.py:
```diff
@@ -26,18 +26,20 @@
 
 class ForTeachersModule(WebsiteModule):
-    def __init__(self, db: Database, achievements: Achievements):
+    def __init__(self, db: Database, achievements: Achievements, surveys: SurveysModule):
```
Great

hasan-sh commented on website/for_teachers.py:
```diff
@@ -297,6 +310,29 @@ def get_class_customization_page(self, user, class_id):
                 class_id=class_id
             ))
 
+    def class_survey(self, class_id):
```
great now:)

hasan-sh commented on utils.py:
```diff
@@ -359,3 +359,23 @@ def can_edit_class(user, Class):
             for second_teacher in Class.get("second_teachers", []))
+
+
+def get_unanswered_questions(survey, new_questions):
```
amazing job!

hasan-sh commented on translations/en/LC_MESSAGES/messages.po:
```diff
@@ -376,6 +376,26 @@ msgstr "Please enter the name of the new class"
+msgid "class_survey_later"
+msgstr "Remind me tomorrow"
+
+msgid "class_survey_question1"
```
ok different questions are fine for now

hasan-sh commented on website/for_teachers.py:
```diff
@@ -297,6 +310,29 @@ def get_class_customization_page(self, user, class_id):
+        questions.append(gettext("class_survey_question2"))
+        questions.append(gettext("class_survey_question3"))
+        questions.append(gettext("class_survey_question4"))
+        unanswered_questions, translate_db = utils.get_unanswered_questions(survey, questions)
```
makes more sense from utils;)

hasan-sh reviewed (commented):
Tested it with the latest changes. I removed the style property since we center the parent with flex, check it out @Annelein ! Will approve there after!

Felienne commented:
Wow that was a lot of hassle @Annelein!! Thanks for being so patient!!

Annelein commented:
> Wow that was a lot of hassle @Annelein!! Thanks for being so patient!!

Ofcourse! Always on the first pull request, getting used to the whole environment :))
